src/backend/distributed/operations/citus_tools.c

```diff
@@ -52,14 +57,61 @@ static void ExecuteCommandsAndStoreResults(StringInfo *nodeNameArray,
 										   bool *statusArray,
 										   StringInfo *resultStringArray,
 										   int commandCount);
-static bool ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort,
-										char *queryString, StringInfo queryResult);
+static bool ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort, const char *user,
+										const char *database, char *queryString,
+										StringInfo queryResultString,
+										bool forceNewConnection);
 static Tuplestorestate * CreateTupleStore(TupleDesc tupleDescriptor,
 										  StringInfo *nodeNameArray, int *nodePortArray,
 										  bool *statusArray,
 										  StringInfo *resultArray, int commandCount);
 
 
+/*
+ * check_connection_to_node sends a simple query from a worker node to another
+ * node, and returns success status.
+ */
+Datum
+check_connection_to_node(PG_FUNCTION_ARGS)
+{
+	char *nodeName = PG_GETARG_TEXT_TO_CSTRING(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	char *userName = PG_GETARG_TEXT_TO_CSTRING_OR_NULL(2);
+	char *databaseName = PG_GETARG_TEXT_TO_CSTRING_OR_NULL(3);
+
+	bool success = CheckConnectionToNode(nodeName, nodePort, userName, databaseName);
+	PG_RETURN_BOOL(success);
+}
+
+
+/*
+ * CheckConnectionToNode sends a simple query from a node to another and returns success status
+ */
+static bool
+CheckConnectionToNode(char *nodeName, int32 nodePort, const char *userName,
+					  const char *databaseName)
+{
+	StringInfo queryResultString = makeStringInfo();
+	bool forceNewConnection = false;
+	bool success = false;
```

Comment:
`sucess` should be declared volatile, see the comments in PG_TRY/PG_CATCH

Also, needs comment why we have pg_try/pg_catch